### ISSUE TITLE
[DISCUSSION] Restructuring HEMCO interface to models

### ISSUE THREAD

**Issue by jimmielin** [geoschem/HEMCO#12]
## Overview
This is a discussion regarding building the interface of HEMCO to external models.

## Current implementation
Current implementation of HEMCO in GEOS-Chem, based on the summary by Melissa @msulprizio, is as follows:

* HEMCO is initialized through an interface **in GEOS-Chem** called `HCOI_GC_Init`, which initializes `HcoConfig`, `HcoState` and calls `HCO_Init`, `HCOX_Init`
* HEMCO is called through the interface `HCOI_GC_Run` which passes data to HEMCO for extensions and calls `HCO_Run`, `HCOX_Run`, and updates diagnostics
* Emission tendencies are applied in `Do_Mixing` (`mixing_mod`) which calls `Do_Tend` to apply tendencies. This calls HEMCO directly through `GetHcoVal`, which obtains emission fluxes and dry deposition frequencies. The fluxes are applied to the species concentration array `State_Chm%Species`

(@msulprizio your slides say this is the case for GC-Classic; are there any differences in GCHP? I believe it calls either `Do_Mixing` or `Do_Tend` in `GIGC_Chunk_Mod.F90`. Maybe it differs in GEOS-5 and GCHP?)

## Proposed restructuring
We should have a unified HEMCO init-run-final interface that serves as a generic interface that accepts information needed from the underlying model, such as
* Model grid information (based on a new `HCO_Grid` derived type object described in #11)
* Met fields
* Restart information (is this needed?)
And it should return fluxes.

I suggest this could be combined with the "Data Broker" component, so this IRF module would be the new thing holding `HcoState`/`ExtState` and it would manage HEMCO for the underlying model.

## Proposed contents of the new IRF interface + Data broker
* Holds the global `HcoState` and `ExtState`
* Holds **arrays** of fluxes computed for given time /grid, identified by an ID - this is so that multiple grids can be handled in the future
* Holds brackets (left/right) of the data needed to compute emissions (See Question 3)

* Routines for init, run and finalizing HEMCO?
* Routines to accept a met field specification (see Question 1 below)
* Routines to receive the fluxes computed for a given domain (passing in grid ID and time); if the fluxes do not exist, then HEMCO is executed auto-magically (this may overlap with the Run routine in the point above. See Question 2)

## Questions

1. How does HEMCO receive met fields? Is it through an `ExtState`? How does it know what met fields are necessary, or it is hard-coded?
2. Should we have an explicit HEMCO "Run" routine, or HEMCO should be automatically ran forward in time when the requested time for emissions is not covered by left/right brackets?
3. How are L/R brackets or time interpolation handled in GEOS-Chem Classic right now? Can we replicate this functionality without breaking GCHP, or avoiding to reinvent the wheel as much as necessary?
4. Are dry deposition fluxes stored in HEMCO too? Can we avoid this and store GEOS-Chem specific fluxes somewhere else, and in tendencies apply dry deposition fluxes + fluxes received from HEMCO in the new IRF module? Having fluxes flow from the model to HEMCO seems kind of scary ... (but it may be useful!)

## Action items
- [ ] Identifying needs and locations where GEOS-Chem directly calls HEMCO
- [ ] Design the new IRF interface
- [ ] Code the new IRF interface & data broker
- [ ] ... potentially testing it with a "dummy model" that will become the new HEMCO Standalone
- [ ] Migrate GEOS-Chem Classic to use the new IRF interface + Data broker
- [ ] ... without breaking GCHP

**Comment by jimmielin** [comment]
Data flow from GEOS-Chem to HEMCO right now is performed through `ExtState` (not sure if it updates stuff in `HcoState` too after initialization of the grid data).

The relevant routines are:
* `ExtState_InitTargets`
* `ExtState_SetFields`
* `ExtState_UpdateFields`, which are updated at every ts.

For the ESMF component this is performed through `HCOI_ESMF_MOD` in Interfaces where `HCO_SetExtState_ESMF` is called w/ `HCO_Imp2Ext2S` gets data from the import state. Changing this may require coordination with Christoph.

The targets are hardcoded, e.g. `ExtState%FLASH_DENS` and defined in `Extensions/hcox_state_mod.F90`:

```fortran
     !----------------------------------------------------------------------
     ! Met fields
     !----------------------------------------------------------------------
     TYPE(ExtDat_2R),  POINTER :: U10M        ! E/W 10m wind speed [m/s]
     TYPE(ExtDat_2R),  POINTER :: V10M        ! N/S 10m wind speed [m/s]
     TYPE(ExtDat_2R),  POINTER :: ALBD        ! Surface albedo [-]
```

These need to be ideally put on the HEMCO grid, so regridding is needed. I suggest moving the whole asserting and allocating HEMCO arrays, such as

Physical copy:
```fortran
    ! SPHU: GEOS-Chem SPHU is in g/kg, while HEMCO uses kg/kg.
    ! NOTE: HEMCO only uses SPHU surface values.
    IF ( ExtState%SPHU%DoUse ) THEN
       CALL HCO_ArrAssert( ExtState%SPHU%Arr, State_Grid%NX, State_Grid%NY, &
                           1, HMRC )

       ! Trap potential errors
       IF ( HMRC /= HCO_SUCCESS ) THEN
          RC     = HMRC
          ErrMsg = 'Error encountered in "HCO_ArrAssert( SPHU )"!'
          CALL GC_Error( ErrMsg, RC, ThisLoc, Instr )
          CALL Flush( HcoState%Config%Err%Lun )
          RETURN
       ENDIF
    ENDIF
```
(in InitTargets)

or directly set:
```fortran
    ! CONV_DEPTH: convective cloud depth [m]
    IF ( ExtState%CONV_DEPTH%DoUse ) THEN
       ExtState%CONV_DEPTH%Arr%Val = State_Met%CONV_DEPTH
    ENDIF
```
(in `UpdateFields`; some first-timestep needed data is set on `SetFields`)

into the new `IntData` module, which will accept met data (potentially with a grid descriptor, in the type of `HcoGrid`), do the regridding, and save it to `ExtState`. The interface to models does not have to touch `ExtState` directly, but instead pass it to the data broker. I will design the interfaces shortly.

**Comment by jimmielin** [comment]
I will be working on this issue and documenting the changes. Most changes will happen within GEOS-Chem's `dev/13.0.0` branch as most of the code is GEOS-Chem specific. The model interface will be documented at the GEOS-Chem wiki at: http://wiki.seas.harvard.edu/geos-chem/index.php/Coupling_HEMCO_with_other_models

Some functionality will be moved to within HEMCO, including the introduction of a new module which contains generic tools for interfacing with HEMCO (e.g. `GetHcoVal`), named `Interfaces/hco_interface_common`.

The GEOS-Chem interface will still live in `GeosCore/hco_interface_gc_mod` and a new GEOS-Chem utilities `GeosCore/hco_utilities_gc_mod` for retrieving met fields and the like.

**Comment by msulprizio** [comment]
@jimmielin Is this work now included in HEMCO 3.0.0? Can we now close this discussion?

**Comment by jimmielin** [comment]
Hi @msulprizio yes we've restructured this partly in GEOS-Chem 13.0.0 + HEMCO 3.0.0, then the next batch of updates is also ready but they're bundled with the HEMCO Grid functionality. We can continue discussion in the HEMCO Grid PR with GEOS-Chem (https://github.com/geoschem/geos-chem/pull/364). I'll close this now!
